[dynamic shapes] avoid unnecessary slices

pianpwk opened the pull request:
Fixes #157289, by extending optimization to slices where the end index exceeds the size.

angelayi commented on test/export/test_export.py:
```diff
@@ -2708,6 +2708,36 @@ def forward(self, x, y, z):
+                mask_length = fill_value.shape[-1]
+                causal_mask[:, :, :, :mask_length] = fill_value
+                return causal_mask
+
```
what does the retraced graph look like?

pianpwk replied:
sym_size_int_4 goes to _1 or _5, or something else

pianpwk committed:
[dynamic shapes] avoid unnecessary slices (#157528)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/157289, by extending optimization to slices where the end index exceeds the size.


Reviewed By: angelayi

Differential Revision: D77745298

Pulled By: pianpwk

huydhn commented:
@pytorchbot merge -i
